From 47cf31a14836f57616b404fbf63c897b236d48f9 Mon Sep 17 00:00:00 2001
From: Matt Madison <matt@madison.systems>
Date: Thu, 19 Nov 2020 07:33:02 -0800
Subject: [PATCH] Add machine ID from fuses to kernel command line

Signed-off-by: Matt Madison <matt@madison.systems>
---
 .../lib/linuxboot/t194/linuxboot_helper.c     | 30 +++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/bootloader/partner/t19x/common/lib/linuxboot/t194/linuxboot_helper.c b/bootloader/partner/t19x/common/lib/linuxboot/t194/linuxboot_helper.c
index 65884cd..3f79dea 100644
--- a/bootloader/partner/t19x/common/lib/linuxboot/t194/linuxboot_helper.c
+++ b/bootloader/partner/t19x/common/lib/linuxboot/t194/linuxboot_helper.c
@@ -530,6 +530,35 @@ done:
 #endif /* CONFIG_ENABLE_SYSTEM_AS_ROOT */
 #endif
 
+static int add_machine_id(char *cmdline, int len, char *param, void *priv)
+{
+	tegrabl_error_t err;
+	uint32_t val;
+	char machine_id[65];
+	int midlen, i;
+
+	if (cmdline == NULL | param == NULL)
+		return -1;
+
+	for (midlen = 0, i = 0; i < 4; i += 1, midlen += 8) {
+		err = tegrabl_fuse_read(FUSE_RESERVED_ODM0 + i, &val, sizeof(val));
+		if (err != TEGRABL_NO_ERROR) {
+			pr_error("Failed to read fuse RESERVED_ODM%d\n", i);
+			return -1;
+		}
+		if (val == 0) {
+			pr_info("No machine ID set in fuses\n");
+			return -1;
+		}
+		if (tegrabl_snprintf(machine_id+midlen, sizeof(machine_id)-midlen, "%08x", val) != 8) {
+			pr_error("Error formatting machine ID word %d\n", i);
+			return -1;
+		}
+	}
+	machine_id[midlen] = '\0';
+	return tegrabl_snprintf(cmdline, len, "%s=%s", param, machine_id);
+}
+
 static struct tegrabl_linuxboot_param extra_params[] = {
 	{ "tegraid", add_tegraid, NULL },
 	{ "maxcpus", add_maxcpus, NULL },
@@ -567,6 +596,7 @@ static struct tegrabl_linuxboot_param extra_params[] = {
 #if defined(CONFIG_ENABLE_NVDEC)
 	{ "nvdec_enabled", tegrabl_linuxboot_add_nvdec_enabled_info, NULL },
 #endif
+	{ "systemd.machine_id", add_machine_id, NULL },
 	{ NULL, NULL, NULL},
 };
 
